### ISSUE TITLE
time to live of DefaultLockRepository has no effect for 2 threads in the same java process when using JdbcLockRegistry

### ISSUE THREAD

**Issue by cuipengfei** [spring-projects/spring-integration#8687]
**In what version(s) of Spring Integration are you seeing this issue?**

5.5.18

**Describe the bug**

time to live of DefaultLockRepository has no effect for 2 threads in the same java process when using JdbcLockRegistry

**To Reproduce**

The following code can reproduce the issue

```java
@Configuration
public class DistributedLock {

    public static final int TIME_TO_LIVE = 30 * 1000;

    @Bean
    public LockRepository defaultLockRepository(DataSource dataSource, @Value(value = "${spring.jpa.properties.hibernate.default_schema}") String schema) {
        DefaultLockRepository defaultLockRepository = new DefaultLockRepository(dataSource);
        defaultLockRepository.setPrefix(schema + ".INT_");
        defaultLockRepository.setTimeToLive(TIME_TO_LIVE);
        return defaultLockRepository;
    }

    @Bean
    public JdbcLockRegistry jdbcLockRegistry(LockRepository lockRepository) {
        return new JdbcLockRegistry(lockRepository);
    }
}
```

```java
@SpringBootApplication
public class SpringExampleApplication implements CommandLineRunner {
    Logger logger = LoggerFactory.getLogger(SpringExampleApplication.class);
    ExecutorService executor = Executors.newFixedThreadPool(10);

    @Autowired
    LockRegistry registry;

    public static void main(String[] args) {
        SpringApplication.run(SpringExampleApplication.class, args);
    }

    @Override
    public void run(String... args) throws Exception {
        runBoth();

//        executor.execute(worker1());
//        executor.execute(worker2());

        // when running both workers in one java process, ttl has no effect
        // when running 2 workers in 2 separate java processes, then ttl works as expected
    }

    private void runBoth() throws InterruptedException {
        logger.info("main - going to run worker 1");
        executor.execute(worker1());

        Thread.sleep(1000); // make sure worker 2 is later enough

        logger.info("main - going to run worker 2");
        executor.execute(worker2());
    }

    private Runnable worker1() {
        return () -> runWorker("worker 1");
    }

    private Runnable worker2() {
        return () -> runWorker("worker 2");
    }

    private void runWorker(String workerName) {
        logger.info(workerName + " - start");

        logger.info(workerName + " - going to lock");

        boolean locked = tryGetLock();
        if (locked) {
            logger.info(workerName + " - lock get ok");
        } else {
            logger.info(workerName + " - lock get fail");
        }

        logger.info(workerName + " - end");
    }

    private boolean tryGetLock() {
        Lock lock = registry.obtain("string123");
        try {
            return lock.tryLock(60, TimeUnit.SECONDS);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
    }
}
```

[![enter image description here][1]][1]

when running 2 workers in 2 separate java processes, TTL works as expected ↑

[![enter image description here][2]][2]

  [1]: https://i.stack.imgur.com/2aPaE.png
  [2]: https://i.stack.imgur.com/HTjf7.png

when running both workers in one java process, TTL has no effect ↑

what is the proper way to make TTL work for 2 threads in same java process as well?

**Expected behavior**

TTL should work the same way regardless of threads running in the same java process or different java processes.

**Sample**

code inline above

**Comment by artembilan** [comment]
I wander what drove you to raise the same concern in different places: https://stackoverflow.com/questions/76755731/time-to-live-of-defaultlockrepository-has-no-effect-for-2-threads-in-the-same-ja.

I'll investigate this in a bit, but for future, please, really raise first on StackOveflow if you are not sure that it is indeed a bug.

Thanks

**Comment by artembilan** [comment]
If I understand your correctly, for failed scenario you run both workers in the same thread.
According to Java `Lock` specification, you can get the lock second time in the same thread. 
Independently of the lock implementation, we honor that specification and allow to re-lock second time in the same thread:
```
		public void lock() {
			this.delegate.lock();
			while (true) {
				try {
					while (!doLock()) {
						Thread.sleep(this.idleBetweenTries.toMillis());
					}
					break;
				}
```

Where a `delegate` is a regular `ReentrantLock`. So, if all requirements are met for it, we are good to update the record in DB since we are in the same thread and work with the same repository.

Would you mind clarifying why your expectations are different?

**Comment by cuipengfei** [comment]
> If I understand your correctly, for failed scenario you run both workers in the same thread. According to Java `Lock` specification, you can get the lock second time in the same thread. Independently of the lock implementation, we honor that specification and allow to re-lock second time in the same thread:
> 
> ```
> 		public void lock() {
> 			this.delegate.lock();
> 			while (true) {
> 				try {
> 					while (!doLock()) {
> 						Thread.sleep(this.idleBetweenTries.toMillis());
> 					}
> 					break;
> 				}
> ```
> 
> Where a `delegate` is a regular `ReentrantLock`. So, if all requirements are met for it, we are good to update the record in DB since we are in the same thread and work with the same repository.
> 
> Would you mind clarifying why your expectations are different?

Thanks for looking into it.

Not the same thread, it was 2 threads/workers in the same java process vs 2 workers each running in its own java process.

**Comment by artembilan** [comment]
Thank you for correcting me!
Well, that behavior is also expected. When you talk about one process (and even different threads ) , you still ask same repository instance and second request is aware about the first. When we deal with several repository instances (different processes in your meaning), they are not aware of each other and we can avoid dead lock only via TTL (expire) property . This is typical approach in distributed systems to let other instances to take the data when locked instance might be already dead or crashed , so it could not unlock. 
This way we recommend a TTL long enough to let the logic in program to be performed, but so reasonable that other instances won’t park for nothing too long.

In other words, the time to live is, essentially , an emergency procedure : the locking process might be still running when we take the lock in other process because of expiration. Therefore it has to be a bit longer than we keep the lock in our program.
Does it make sense?

**Comment by cuipengfei** [comment]
> 

Thanks, that does make sense.

The issue I was facing is like this:

**request 1 hits server 1**, while handling the request, the **thread** **sometimes** **hangs**(due to a weird issue of a 3rd party jar which we can not replace)
**when** that **hanging** happens, our code **won't have a chance to release the lock**

then **request 2** comes, if it hits **other servers**, then **it's fine**. but **if it hits server 1**, then **request 2** won't be able to get the lock since the hanging thread **never released the lock** and **TTL does not help in this case**.

In **summary**:
In **one java process**, **a thread gets the lock** but somehow due to bugs or whatever reason it **did not get a chance to  release** the lock. 
Then **subsequent threads in this java process** **won't be able to obtain the same lock** and can not proceed with unfinished job.
**In this case, are there any recommended ways to allow the subsequent threads to be able to get the lock?**

**Comment by artembilan** [comment]
Well, this is really the problem of that part of your program where you have to break that dead lock somehow, but that's already out of distributed lock implementation.
You can look into a `JdbcLockRegistry.expireUnusedOlderThan()` though, but just imaging what would you do if you wouldn't use this `JdbcLockRegistry`, but regular `ReentrantLock` since you really talk about a single JVM problem.

So, according to you latest explanation for the problem I treat this as `Works as Designed` and closing.

It is always not good to keep blocked threads: you need to come up with the solution where you can interrupt that dead lock then you can use a `lockInterruptibly()` from that `JdbcLockRegistry`.
